dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLlvm.java

```diff
@@ -1,18 +1,1434 @@
+        for (final TopLevelEntityContext entity : ctx.topLevelEntity()) {
+            if (entity.funcDecl() != null) {
+                visitFuncHeader(entity.funcDecl().funcHeader());
+            }
+            if (entity.funcDef() != null) {
+                visitFuncHeader(entity.funcDef().funcHeader());
+            }
+            // FIXME: Declare global variables
+        }
+        for (final TopLevelEntityContext entity : ctx.topLevelEntity()) {
+            //FIXME: We need to parse the declaration of globals (similar to func headers) first
+            // because the initializer may refer to other globals.
+            if (entity.globalDecl() != null || entity.globalDef() != null || entity.typeDef() != null) {
+                entity.accept(this);
+            }
+        }
+
+        // Parse definitions
+        for (final TopLevelEntityContext entity : ctx.topLevelEntity()) {
+            if (entity.metadataDef() == null &&
+                    entity.globalDecl() == null &&
+                    entity.globalDef() == null &&
+                    entity.typeDef() == null &&
+                    entity.funcDecl() == null) {
+                entity.accept(this);
+            }
+        }
+        return null;
+    }
+
+    // ----------------------------------------------------------------------------------------------------------------
+    // Top Level Entities
+
+    @Override
+    public Expression visitTypeDef(TypeDefContext ctx) {
+        final String name = localIdent(ctx.LocalIdent());
+        typeDefinitionMap.put(name, ctx);
+        return null;
+    }
+
+    @Override
+    public Expression visitFuncHeader(FuncHeaderContext ctx) {
+        final String name = globalIdent(ctx.GlobalIdent());
+        final Type returnType = parseType(ctx.type());
+        final List<Type> parameterTypes = new ArrayList<>();
+        final List<String> parameterNames = new ArrayList<>();
+        int unnamedParameters = 0;
+        for (ParamContext parameter : ctx.params().param()) {
+            if (parameter.type().getText().equals("metadata")) {
+                //TODO also declare this.
+                return null;
+            }
+            parameterTypes.add(parseType(parameter.type()));
+            final String parameterName = parameter.LocalIdent() != null ?
+                    localIdent(parameter.LocalIdent()) :
+                    "dummy_" + (unnamedParameters++);
+            parameterNames.add(adjustRegisterName(parameterName));
+        }
+        final boolean isVarArgs = ctx.params().ellipsis != null;
+        final FunctionType functionType = types.getFunctionType(returnType, parameterTypes, isVarArgs);
+        //TODO there are more attributes in ctx
+        final var declaredFunction = new Function(name, functionType, parameterNames, functionCounter++, null);
+
+        program.addFunction(declaredFunction);
+        constantMap.put(name, declaredFunction);
```

Comment:
It doesn't interfere at all. The map is used to replace references to function names by the concrete function.
If a call goes to a generic function pointer, than it will obviously not get replaced.
Some pass in the processing pipeline resolves calls to function pointers to direct calls, but that happens outside of the parser/visitor.